### PR TITLE
Updated for Jetpack Compose dev15 and Android Studio 4.2 Canary 5

### PR DESCRIPTION
**Pull request by madhavajay**
I got everything working, however I can't figure out the correct way to handle the three unchecked casts properly.
I left the @Suppress("UNCHECKED_CAST") commented near each one.

I guess this is something to do with Covariance / Contravariance etc, but I dont know enough about how to solve it. Seems theres a List<Any> and a Map with BackStack<*>. Would one solution be to have some kind of known Routing interface or class which is subclassed or implemented so that the type and its sub/super are known?

I would be interested in knowing how this should be fixed so that I can understand this whole <in T> <out T> <*> thing in Kotlin.

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -1,4 +1,4 @@
 include ':app-nested-containers'
 include ':app-lifelike'
 include ':router'
-rootProject.name = "compose-router"
+rootProject.name = "compose-router"
```
Comment (ShikaSD) [inline settings.gradle]: Seems like you removed the new line at the end of the file

File: router/build.gradle [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
-apply plugin: 'kotlin-android'
-apply plugin: 'kotlin-android-extensions'
+plugins {
+    id 'com.android.library'
+    id 'kotlin-android'
+    id 'kotlin-android-extensions'
+    id 'kotlin-kapt'
+}
+
+def compose_version = "0.1.0-dev15"
```
Comment (ShikaSD) [inline router/build.gradle]: Instead of declaring it on in each file separately, you could add it to root `build.gradle` next to `kotlin_version`:
```
ext {
    compose_version = "0.1.0-dev15"
}
```
Now it is going to be available in all modules implicitly.

File: router/build.gradle [hunk 3]
```diff
@@ -26,32 +30,31 @@ android {
...
+    kotlinOptions {
+        jvmTarget = "1.8"
+        freeCompilerArgs += ["-Xallow-jvm-ir-dependencies", "-Xskip-prerelease-check"]
     }
 }
 
 dependencies {
-    implementation fileTree(dir: 'libs', include: ['*.jar'])
-    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    implementation "org.jetbrains.kotlin:kotlin-stdlib:$project.ext.kotlin_version"
```
Comment (ShikaSD) [inline router/build.gradle]: Why the change to `project.ext`? Its fields are exported automatically :)

File: router/build.gradle [hunk 4]
```diff
@@ -26,32 +30,31 @@ android {
...
-    implementation 'androidx.core:core-ktx:1.3.0'
-    implementation 'androidx.ui:ui-core:0.1.0-dev14'
-    implementation 'androidx.ui:ui-layout:0.1.0-dev14'
-    implementation 'androidx.ui:ui-material:0.1.0-dev14'
-    implementation 'androidx.ui:ui-tooling:0.1.0-dev14'
-    implementation 'androidx.ui:ui-foundation:0.1.0-dev14'
-    testImplementation 'junit:junit:4.12'
+    implementation 'androidx.core:core-ktx:1.3.1'
+
+    implementation 'androidx.compose.ui:ui:0.1.0-dev15'
```
Comment (ShikaSD) [inline router/build.gradle]: Maybe reuse `compose_version` here and lower?
E.g. `implementation "androidx.compose.ui:ui:$compose_version"`. Also notice double quotes instead of single ones, it allows to use these placeholders in string.

File: build.gradle [hunk 5]
```diff
@@ -1,28 +1,36 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
-    buildscript {
-        ext.kotlin_version = "1.3.72"
-        repositories {
-            google()
-            jcenter()
-
+buildscript {
+    ext {
+        compose_version = '0.1.0-dev13'
```
Comment (ShikaSD) [inline build.gradle]: Why this one is different from dev-15?
Also you seem to not use this variable

File: build.gradle [hunk 6]
```diff
@@ -1,28 +1,36 @@
...
+    }
+    ext.kotlin_version = "1.4-M3"
```
Comment (ShikaSD) [inline build.gradle]: You can combine ext blocks like this
```
ext {
    compose_version = '0.1.0-dev15'
    kotlin_version = '1.4-M3'
}

File: app-nested-containers/src/main/java/com/example/nestedcontainers/composable/SomeChild.kt [hunk 7]
```diff
@@ -165,25 +164,22 @@ interface SomeChild {
...
             val textStyle = MaterialTheme.typography.body1
 
             Surface(color = colorResource(bgColor)) {
                 Column(modifier = Modifier.padding(16.dp)) {
-                    Box(modifier = Modifier.wrapContentHeight(Alignment.CenterVertically)) {
-                        Button(onClick = onButtonClick, modifier = Modifier.ripple(true)) {
-                            Text("$name.NEXT", style = textStyle.copy(color = Color.White))
-                        }
+                    Box(modifier = Modifier.wrapContentHeight(Alignment.CenterVertically).clickable(onClick = onButtonClick)) {
+
```
Comment (ShikaSD) [inline app-nested-containers/src/main/java/com/example/nestedcontainers/composable/SomeChild.kt:179]: Extra line here:)

File: app-nested-containers/build.gradle [hunk 8]
```diff
@@ -1,11 +1,15 @@
...
-apply plugin: 'kotlin-android'
-apply plugin: 'kotlin-android-extensions'
+plugins {
+    id 'com.android.application'
+    id 'kotlin-android'
+    id 'kotlin-android-extensions'
+    id 'kotlin-kapt'
+}
+
+def compose_version = "0.1.0-dev15"
```
Comment (ShikaSD) [inline app-nested-containers/build.gradle]: Same thing as in other file, maybe define it once for the whole project?

File: app-nested-containers/build.gradle [hunk 9]
```diff
@@ -26,31 +30,32 @@ android {
...
     implementation 'androidx.appcompat:appcompat:1.1.0'
-    implementation 'androidx.core:core-ktx:1.3.0'
-    implementation 'androidx.ui:ui-core:0.1.0-dev14'
-    implementation 'androidx.ui:ui-layout:0.1.0-dev14'
-    implementation 'androidx.ui:ui-material:0.1.0-dev14'
-    implementation 'androidx.ui:ui-tooling:0.1.0-dev14'
-    implementation 'androidx.ui:ui-foundation:0.1.0-dev14'
-    //implementation 'com.github.zsoltk:compose-router:0.11.1'
+    implementation 'androidx.core:core-ktx:1.3.1'
+    implementation 'androidx.compose.ui:ui:0.1.0-dev15'
```
Comment (ShikaSD) [inline app-nested-containers/build.gradle]: Same thing as in other files, maybe reuse `compose_version` variable?

File: app-lifelike/src/main/java/com/example/lifelike/composable/loggedin/AlbumList.kt [hunk 10]
```diff
@@ -43,7 +38,9 @@ interface AlbumList {
             val image = imageResource(R.drawable.placeholder)
             val typography = MaterialTheme.typography
 
-            Clickable(onClick = { onAlbumSelected(album) }, modifier = Modifier.ripple(true)) {
+            Box(
+                    modifier = Modifier.clickable(onClick = { onAlbumSelected(album) })
```
Comment (ShikaSD) [inline app-lifelike/src/main/java/com/example/lifelike/composable/loggedin/AlbumList.kt:44]: Intentation seems off, supposed to be 4 spaces instead of 8 :)

File: app-lifelike/build.gradle [hunk 11]
```diff
@@ -1,15 +1,20 @@
...
+plugins {
+    id 'com.android.application'
+    id 'kotlin-android'
+    id 'kotlin-android-extensions'
+    id 'kotlin-kapt'
+}
 
-android {
-    compileSdkVersion 29
+def compose_version = "0.1.0-dev15"
```
Comment (ShikaSD) [inline app-lifelike/build.gradle]: Same thing as in other files, maybe define this version once for the whole project?

File: app-lifelike/build.gradle [hunk 12]
```diff
@@ -26,31 +31,33 @@ android {
...
+
     implementation 'androidx.appcompat:appcompat:1.1.0'
-    implementation 'androidx.core:core-ktx:1.3.0'
-    implementation 'androidx.ui:ui-core:0.1.0-dev14'
-    implementation 'androidx.ui:ui-layout:0.1.0-dev14'
-    implementation 'androidx.ui:ui-material:0.1.0-dev14'
-    implementation 'androidx.ui:ui-tooling:0.1.0-dev14'
-    implementation 'androidx.ui:ui-foundation:0.1.0-dev14'
+    implementation 'androidx.core:core-ktx:1.3.1'
+    implementation 'androidx.compose.ui:ui:0.1.0-dev15'
```
Comment (ShikaSD) [inline app-lifelike/build.gradle]: Same thing as in other files, maybe use variable to specify version?

File: app-lifelike/build.gradle [hunk 13]
```diff
@@ -26,31 +31,33 @@ android {
...
-        jvmTarget = '1.8'
-    }
     buildFeatures {
         compose true
     }
     composeOptions {
-        kotlinCompilerVersion "1.3.70-dev-withExperimentalGoogleExtensions-20200424"
-        kotlinCompilerExtensionVersion "0.1.0-dev14"
+        kotlinCompilerVersion "1.4.0-dev-withExperimentalGoogleExtensions-20200720"
+        kotlinCompilerExtensionVersion "$compose_version"
```
Comment (ShikaSD) [inline app-lifelike/build.gradle]: You actually can use just a variable directly.
E.g. `compose_version` instead of `"$compose_version"`

File: app-lifelike/src/main/java/com/example/lifelike/composable/loggedin/AlbumList.kt [hunk 14]
```diff
@@ -29,7 +24,7 @@ interface AlbumList {
     companion object {
         @Composable
         fun Content(onAlbumSelected: (Album) -> Unit) {
-            VerticalScroller {
+            ScrollableColumn {
                 Column {
```
Comment (ShikaSD) [inline app-lifelike/src/main/java/com/example/lifelike/composable/loggedin/AlbumList.kt]: Not sure, but do we need `Column` inside `ScrollableColumn` ?

File: build.gradle [hunk 15]
```diff
@@ -1,28 +1,36 @@
...
     }
+}
 
-    allprojects {
-        repositories {
-            google()
-            jcenter()
-            maven { url 'https://jitpack.io' }
+allprojects {
+    ext.kotlin_version = "1.4-M3"
```
Comment (ShikaSD) [inline build.gradle]: Do you need to repeat this definition here?
If you do, maybe define `compose_version` here as well?

File: router/src/main/java/com/github/zsoltk/compose/router/Router.kt [hunk 16]
```diff
@@ -85,13 +88,13 @@ private fun <T> fetchBackStack(key: String, defaultElement: T, override: T?): Ba
     val upstreamBundle = AmbientSavedInstanceState.current
     val onElementRemoved: (Int) -> Unit = { upstreamBundle.remove(key(it)) }
 
+    // @Suppress("UNCHECKED_CAST")
     val existing = backStackMap[key] as BackStack<T>?
-
+    // @Suppress("UNCHECKED_CAST")
```
Comment (ShikaSD) [inline router/src/main/java/com/github/zsoltk/compose/router/Router.kt]: Either uncomment or remove these comments :)
I think both variants are fine, as we don't have "warnings as errors" policy for now.

File: router/src/main/java/com/github/zsoltk/compose/router/Router.kt [hunk 17]
```diff
@@ -85,13 +88,13 @@ private fun <T> fetchBackStack(key: String, defaultElement: T, override: T?): Ba
...
     return when {
-        override != null -> BackStack(override, onElementRemoved)
+        override != null -> BackStack(override as T, onElementRemoved)
```
Comment (ShikaSD) [inline router/src/main/java/com/github/zsoltk/compose/router/Router.kt:92]: Seems like this cast is not needed?

**Review by ShikaSD — CHANGES_REQUESTED** [review]
Hey, amazing job of converting these.
I have left a couple of comments, mostly referring to build files.

Regarding unchecked type checks, seems like for now we cannot avoid them all, as we are converting between more broad `BackStack<Any>` to very specific `BackStack<T>`. I think in these cases it is fine anyway, as we know that types are consistent based on the logic of the lib :)

File: router/build.gradle (hunk 3 above)
Comment (madhavajay) [reply router/build.gradle]: Sorry, I was trying to get it to compile, and thought it was a dependency error, when it was really my fault for changing the core module plugin to application instead of library which resulted in an error which wasn't super clear.
Do you want me to just revert this whole change? If the code compiles and runs without the first 2 lines why are they still needed?

File: router/src/main/java/com/github/zsoltk/compose/router/Router.kt (hunk 17 above)
Comment (madhavajay) [reply router/src/main/java/com/github/zsoltk/compose/router/Router.kt:92]: Not sure why my Android Studio complains when I remove that:
https://imgur.com/a/YGVjvru

Comment (ShikaSD) [reply router/src/main/java/com/github/zsoltk/compose/router/Router.kt:92]: Doesn't complain for me 🤔 
Is it compiling without `as` though?

File: router/build.gradle (hunk 3 above)
Comment (ShikaSD) [reply router/build.gradle]: You still need stdlib for sure, but first one can be removed, yes :)

File: app-lifelike/src/main/java/com/example/lifelike/composable/LoggedOut.kt [hunk 18]
```diff
@@ -1,11 +1,7 @@
 package com.example.lifelike.composable
 
 import androidx.compose.Composable
-import com.example.lifelike.composable.loggedout.RegConfirmSmsCode
-import com.example.lifelike.composable.loggedout.RegFinal
-import com.example.lifelike.composable.loggedout.RegUserName
-import com.example.lifelike.composable.loggedout.RegUserPhone
-import com.example.lifelike.composable.loggedout.Splash
+import com.example.lifelike.composable.loggedout.*
```
Comment (ShikaSD) [inline app-lifelike/src/main/java/com/example/lifelike/composable/LoggedOut.kt]: Star imports are considered a bad practice, because you can accidentally import something not needed from a package.
Can you please revert it?

You can also update studio to avoid these changes in `Preferences | Editor | Code Style | Kotlin | Imports` and select `Use single name import` everywhere :)

File: router/src/main/java/com/github/zsoltk/compose/router/Router.kt (hunk 17 above)
Comment (ShikaSD) [reply router/src/main/java/com/github/zsoltk/compose/router/Router.kt:92]: Ah, needed to update Kotlin plugin. Seems like a bug in type inference, let's leave `as T` for now :)

File: .github/workflows/gradle.yml [hunk 19]
```diff
@@ -0,0 +1,41 @@
+name: Build Gradle Tasks
```
Comment (ShikaSD) [inline .github/workflows/gradle.yml]: Thanks for contributing that, but can you open another PR to add the Actions config and the badge?
It would be easier to process and discuss as a separate change and focus only on migrating to dev-15 here :)

**Comment by madhavajay** [comment]
Okay, I have made all the changes and squashed into a single commit. I will open a separate PR for the CI stuff after this is merged and I can rebase. Let me know if theres anything else you need me to change. 😊

**Comment by ShikaSD** [comment]
@madhavajay amazing, thanks 👐 . I still see a new yaml file in the diff, can you recheck if you have removed everything?
Otherwise, LGTM.
cc: @zsoltk

**Comment by madhavajay** [comment]
@ShikaSD Sorry, I missed that when I squashed the commits. It's gone now, so I think we're ready for merge. Thanks for the patience.

**Comment by zsoltk** [comment]
@madhavajay @ShikaSD thank you both!
